source/draw/gpu/opengl/backend/gles3/Gles3ShaderProgram.ooc

```diff
@@ -31,62 +31,62 @@ Gles3ShaderProgram: class extends GLShaderProgram {
-		version(debugGL) { validateEnd("ShaderProgram setUniform~IntArray name:%s" format(name)) }
+		version(debugGL) { validateEnd("ShaderProgram setUniform~IntArray name:%s" format(name), true) }
 	}
 	setUniform: override func ~Float (name: String, x: Float) {
 		version(debugGL) { validateStart("ShaderProgram setUniform~Float") }
 		glUniform1f(glGetUniformLocation(this _backend, name), x)
-		version(debugGL) { validateEnd("ShaderProgram setUniform~Float name:%s x=%f" format(name, x)) }
+		version(debugGL) { validateEnd("ShaderProgram setUniform~Float name:%s x=%f" format(name, x), true) }
 	}
 	setUniform: override func ~Float2 (name: String, x, y: Float) {
 		version(debugGL) { validateStart("ShaderProgram setUniform~Float2") }
 		glUniform2f(glGetUniformLocation(this _backend, name), x, y)
-		version(debugGL) { validateEnd("ShaderProgram setUniform~Float2 name:%s x=%f y=%f" format(name, x, y)) }
+		version(debugGL) { validateEnd("ShaderProgram setUniform~Float2 name:%s x=%f y=%f" format(name, x, y), true) }
 	}
 	setUniform: override func ~Float3 (name: String, x, y, z: Float) {
 		version(debugGL) { validateStart("ShaderProgram setUniform~Float3") }
 		glUniform3f(glGetUniformLocation(this _backend, name), x, y, z)
-		version(debugGL) { validateEnd("ShaderProgram setUniform~Float3 name:%s x=%f y=%f z=%f" format(name, x, y, z)) }
+		version(debugGL) { validateEnd("ShaderProgram setUniform~Float3 name:%s x=%f y=%f z=%f" format(name, x, y, z), true) }
 	}
 	setUniform: override func ~Float4 (name: String, x, y, z, w: Float) {
 		version(debugGL) { validateStart("ShaderProgram setUniform~Float4") }
 		glUniform4f(glGetUniformLocation(this _backend, name), x, y, z, w)
-		version(debugGL) { validateEnd("ShaderProgram setUniform~Float4 name:%s x=%f y=%f z=%f w=%f" format(name, x, y, z, w)) }
+		version(debugGL) { validateEnd("ShaderProgram setUniform~Float4 name:%s x=%f y=%f z=%f w=%f" format(name, x, y, z, w), true) }
 	}
 	setUniform: override func ~FloatArray (name: String, values: Float*, count: Int) {
 		version(debugGL) { validateStart("ShaderProgram setUniform~FloatArray") }
 		glUniform1fv(glGetUniformLocation(this _backend, name), count, values)
-		version(debugGL) { validateEnd("ShaderProgram setUniform~FloatArray name:%s" format(name)) }
+		version(debugGL) { validateEnd("ShaderProgram setUniform~FloatArray name:%s" format(name), true) }
 	}
 	setUniform: override func ~FloatTransform2D (name: String, value: FloatTransform2D) {
 		version(debugGL) { validateStart("ShaderProgram setUniform~FloatTransform2D") }
 		glUniformMatrix3fv(glGetUniformLocation(this _backend, name), 1, 0, value& as Float*)
-		version(debugGL) { validateEnd("ShaderProgram setUniform~FloatTransform2D name:%s value: " + value toString()) }
+		version(debugGL) { strValue := value toString(); validateEnd("ShaderProgram setUniform~FloatTransform2D name:%s value: %s" format(strValue), true); strValue free() }
```

Comment:
Hold it right there. You only have one argument to `format`, and the other could be auto-freed by `&` operator.